File: electron/capture/CaptureManager.ts
/**
 * CaptureManager：编排层
 *
 * 整合 WindowWatcher + Screenshot + CaptureDecision + PrivacyGuard + IncognitoDetector + SegmentRepository。
 *
 * 职责：
 *  - 启动/停止/暂停/恢复全链路捕获
 *  - 监听 CaptureDecision 事件，持久化 Segment 到数据库
 *  - 监听 IncognitoDetector 事件，广播 IPC 通知渲染进程（桌面伙伴遮眼拉帘）
 *  - 管理截图持久化设置（saveScreenshots / retentionDays）
 *  - 单例导出 getCaptureManager()
 *
 * 硬约束：不监听键盘，仅编排窗口/截图/隐私模块。
 */
import { EventEmitter } from 'node:events'
import { powerMonitor } from 'electron'
import type { RecordingState } from '@/types'
import { SegmentRepository } from '../db/repositories/SegmentRepository'
import { WindowWatcher } from './WindowWatcher'
import { Screenshot } from './Screenshot'
import { PrivacyGuard } from './PrivacyGuard'
import { CaptureDecision } from './CaptureDecision'
import type {
  SegmentCreatedPayload,
  SegmentMergedPayload,
  PrivacyPlaceholderPayload,
  StateChangePayload
} from './CaptureDecision'
import { getMainWindow } from '../main/window'

/** IPC 广播通道名：无痕模式检测到 / 清除 */
const IPC_INCOGNITO_DETECTED = 'mascot:incognito-detected'
const IPC_INCOGNITO_CLEARED = 'mascot:incognito-cleared'
/** 与 CaptureDecision 空闲阈值对齐：3 分钟无活动进入 idle */
const SYSTEM_IDLE_THRESHOLD_SECONDS = 3 * 60
/** 系统活动轮询间隔（毫秒） */
const ACTIVITY_POLL_INTERVAL_MS = 15000

/**
 * CaptureManager：捕获全链路编排单例。
 *
 * 事件：
 *  - 'state-change'：记录状态变化，携带 RecordingState
 *  - 'segment-created'：片段已创建并持久化，携带 WorkSegment
 *  - 'segment-merged'：片段已合并
 *  - 'privacy-placeholder'：隐私占位已创建
 */
export class CaptureManager extends EventEmitter {
  private watcher: WindowWatcher
  private screenshot: typeof Screenshot
  private privacyGuard: PrivacyGuard
  private decision: CaptureDecision
  private activityTimer: NodeJS.Timeout | null = null

  /** 基础状态（不含 privacy 覆盖） */
  private baseState: RecordingState = 'idle'
  /** 是否处于隐私模式（无痕窗口激活时） */
  private privacyMode = false

  /** 截图持久化设置 */
  private saveScreenshots = false
  private screenshotRetentionDays = 0
  /**
   * 是否允许活跃窗口截图失败后整屏降级。默认 false（隐私安全）。
   * 由 setAllowFullScreenshotFallback 同步自 SettingsStore，并下发到 CaptureDecision。
   */
  private allowFullScreenshotFallback = false

  constructor() {
    super()
    this.watcher = new WindowWatcher()
    this.screenshot = Screenshot
    this.privacyGuard = new PrivacyGuard()
    this.decision = new CaptureDecision(this.watcher, this.screenshot, this.privacyGuard)
    this.setupEventListeners()
  }

  // ===================== 捕获控制 =====================

  /** 启动全链路捕获 */
  startCapture(): boolean {
    this.privacyGuard.seedDefaultRules()
    this.watcher.start()
    this.decision.start()
    this.setBaseState('recording')
    this.startActivityMonitor()
    // 启动时清理过期截图
    if (this.screenshotRetentionDays > 0) {
      this.screenshot.cleanExpiredScreenshots(this.screenshotRetentionDays)
    }
    return true
  }

  /** 停止捕获 */
  stopCapture(): boolean {
    this.stopActivityMonitor()
    this.decision.stop()
    this.watcher.stop()
    this.setBaseState('idle')
    return true
  }

  /** 暂停捕获 */
  pauseCapture(): boolean {
    this.stopActivityMonitor()
    this.decision.pause()
    this.setBaseState('paused')
    return true
  }

  /** 恢复捕获 */
  resumeCapture(): boolean {
    this.decision.resume()
    this.setBaseState('recording')
    this.startActivityMonitor()
    this.handleUserBecameActive()
    return true
  }

  /** 获取当前记录状态（含 privacy 覆盖） */
  getRecordingState(): RecordingState {
    return this.privacyMode ? 'privacy' : this.baseState
  }

  // ===================== 模块访问 =====================

  getPrivacyGuard(): PrivacyGuard {
    return this.privacyGuard
  }

  getWindowWatcher(): WindowWatcher {
    return this.watcher
  }

  getCaptureDecision(): CaptureDecision {
    return this.decision
  }

  // ===================== 设置 =====================

  /** 设置是否持久保存截图 */
  setSaveScreenshots(enabled: boolean): void {
    this.saveScreenshots = enabled
  }

  /** 设置截图保留天数 */
  setScreenshotRetentionDays(days: number): void {
    this.screenshotRetentionDays = Math.max(0, Math.min(7, days))
  }

  /**
   * 设置是否允许活跃窗口截图失败后整屏降级，并立即下发到 CaptureDecision。
   * 默认 false：窗口截图失败即跳过，绝不自动整屏。
   */
  setAllowFullScreenshotFallback(enabled: boolean): void {
    this.allowFullScreenshotFallback = enabled
    this.decision.setAllowFullScreenshotFallback(enabled)
  }

  /** 查询当前是否允许整屏降级 */
  isFullScreenshotFallbackAllowed(): boolean {
    return this.allowFullScreenshotFallback
  }

  /** 立即清理过期截图 */
  cleanExpiredScreenshots(): void {
    if (this.screenshotRetentionDays > 0) {
      this.screenshot.cleanExpiredScreenshots(this.screenshotRetentionDays)
    }
  }

  // ===================== 事件监听 =====================

  private setupEventListeners(): void {
    // CaptureDecision → 持久化 Segment
    this.decision.on('segment-created', (payload: SegmentCreatedPayload) => {
      this.onSegmentCreated(payload)
    })

    this.decision.on('segment-merged', (payload: SegmentMergedPayload) => {
      SegmentRepository.update(payload.id, payload.patch)
      this.emit('segment-merged', payload)
    })

    this.decision.on('privacy-placeholder', (payload: PrivacyPlaceholderPayload) => {
      const segment = SegmentRepository.insert(payload.segment)
      this.emit('privacy-placeholder', segment)
    })

    this.decision.on('state-change', (payload: StateChangePayload) => {
      // CaptureDecision 的 idle 检测会 emit idle 状态
      if (payload.state === 'idle') {
        this.setBaseState('idle')
      }
    })

    // IncognitoDetector → 广播 IPC + 隐私模式
    const incognitoDetector = this.privacyGuard.getIncognitoDetector()
    incognitoDetector.watch(this.watcher)

    incognitoDetector.on('incognito-detected', () => {
      this.privacyMode = true
      this.broadcastState()
      this.broadcastToRenderer(IPC_INCOGNITO_DETECTED)
    })

    incognitoDetector.on('incognito-cleared', () => {
      this.privacyMode = false
      this.broadcastState()
      this.broadcastToRenderer(IPC_INCOGNITO_CLEARED)
    })
  }

  /** 片段创建回调：持久化到数据库，可选保存截图 */
  private onSegmentCreated(payload: SegmentCreatedPayload): void {
    const segment = SegmentRepository.insert(payload.segment)
    // 整屏降级元数据审计：若本次截图经由整屏降级获得，记录所截取的屏幕范围与 segment id
    if (payload.fallbackDisplayBounds) {
      const b = payload.fallbackDisplayBounds
      console.warn(
        `[CaptureManager] segment ${segment.id} 由整屏降级捕获，` +
          `displayBounds={x:${b.x},y:${b.y},width:${b.width},height:${b.height}} ` +
          `app=${segment.appName} date=${segment.date}`
      )
    }
    // 若开启截图保存，持久化截图到 userData/screenshots/
    if (this.saveScreenshots && payload.screenshotBuffer) {
      const persistentPath = this.screenshot.saveScreenshot(
        payload.screenshotBuffer,
        segment.date,
        segment.id
      )
      if (persistentPath) {
        const updated = SegmentRepository.update(segment.id, { screenshotPath: persistentPath })
        if (updated) {
          this.emit('segment-created', updated)
          return
        }
      }
    }
    this.emit('segment-created', segment)
  }

  // ===================== 内部工具 =====================

  private setBaseState(state: RecordingState): void {
    this.baseState = state
    this.broadcastState()
  }

  private startActivityMonitor(): void {
    if (this.activityTimer) return

    const poll = (): void => {
      if (this.baseState === 'paused') return

      try {
        const idleSeconds = powerMonitor.getSystemIdleTime()

        if (idleSeconds >= SYSTEM_IDLE_THRESHOLD_SECONDS) {
          if (!this.privacyMode && this.baseState !== 'idle') {
            this.setBaseState('idle')
          }
          return
        }

        this.handleUserBecameActive()
      } catch (e) {
        console.warn('[CaptureManager] 读取系统空闲时间失败:', e instanceof Error ? e.message : String(e))
      }
    }

    poll()
    this.activityTimer = setInterval(poll, ACTIVITY_POLL_INTERVAL_MS)
  }

  private stopActivityMonitor(): void {
    if (this.activityTimer) {
      clearInterval(this.activityTimer)
      this.activityTimer = null
    }
  }

  private handleUserBecameActive(): void {
    if (this.baseState === 'paused') return

    const activeWindow = this.watcher.getActiveWindowSnapshot() ?? this.watcher.getLastWindowInfo()
    this.decision.wakeFromActivity(activeWindow)

    if (!this.privacyMode && this.baseState === 'idle') {
      this.setBaseState('recording')
    }
  }

  private broadcastState(): void {
    const state = this.getRecordingState()
    this.emit('state-change', state)
    // 同步通知渲染进程状态变化
    this.broadcastToRenderer('capture:state-changed', state)
  }

  private broadcastToRenderer(channel: string, ...args: unknown[]): void {
    try {
      const win = getMainWindow()
      if (win && !win.isDestroyed()) {
        win.webContents.send(channel, ...args)
      }
    } catch (e) {
      console.warn('[CaptureManager] 广播 IPC 失败:', e instanceof Error ? e.message : String(e))
    }
  }
}

// ===================== 单例 =====================

let managerInstance: CaptureManager | null = null

/** 获取 CaptureManager 单例 */
export function getCaptureManager(): CaptureManager {
  if (!managerInstance) {
    managerInstance = new CaptureManager()
  }
  return managerInstance
}

/** 重置单例（仅供测试） */
export function resetCaptureManager(): void {
  if (managerInstance) {
    managerInstance.stopCapture()
    managerInstance = null
  }
}

File: src-tauri/src/capture/capture_manager.rs
//! CaptureManager：捕获全链路编排单例（对应 electron/capture/CaptureManager.ts）
//!
//! 整合 WindowWatcher + Screenshot + CaptureDecision + PrivacyGuard + IncognitoDetector + SegmentRepository。
//!
//! 职责：
//!  - 启动/停止/暂停/恢复全链路捕获
//!  - 监听 CaptureDecision 事件，持久化 Segment 到数据库
//!  - 监听 IncognitoDetector 事件，广播 IPC 通知渲染进程（桌面伙伴遮眼拉帘）
//!  - 管理截图持久化设置（saveScreenshots / retentionDays）
//!  - 单例导出 get_capture_manager()
//!
//! 硬约束：不监听键盘，仅编排窗口/截图/隐私模块。

use std::sync::Mutex;
use std::time::{SystemTime, UNIX_EPOCH};

use once_cell::sync::Lazy;

use crate::capture::capture_decision::CaptureDecision;
use crate::capture::privacy_guard::PrivacyGuard;
use crate::capture::window_watcher::{WindowInfo, WindowWatcher};
use crate::events::bus::{AppEvent, EventBus};
use crate::models::RecordingState;

/// 与 CaptureDecision 空闲阈值对齐：3 分钟无活动进入 idle
pub const SYSTEM_IDLE_THRESHOLD_SECONDS: u64 = 3 * 60;
/// 系统活动轮询间隔（毫秒）
pub const ACTIVITY_POLL_INTERVAL_MS: u64 = 15000;

/// 捕获状态
#[derive(Debug, Clone)]
pub struct CaptureState {
    /// 基础状态（不含 privacy 覆盖）
    pub base_state: RecordingState,
    /// 是否处于隐私模式（无痕窗口激活时）
    pub privacy_mode: bool,
}

impl CaptureState {
    /// 获取当前记录状态（含 privacy 覆盖）
    pub fn effective_state(&self) -> RecordingState {
        if self.privacy_mode {
            RecordingState::Privacy
        } else {
            self.base_state.clone()
        }
    }
}

impl Default for CaptureState {
    fn default() -> Self {
        Self {
            base_state: RecordingState::Idle,
            privacy_mode: false,
        }
    }
}

/// CaptureManager：捕获全链路编排单例。
pub struct CaptureManager {
    /// 窗口监听器
    watcher: Mutex<Option<WindowWatcher>>,
    /// 隐私守卫
    privacy_guard: PrivacyGuard,
    /// 截图决策
    decision: CaptureDecision,
    /// 捕获状态
    state: Mutex<CaptureState>,
    /// 是否持久保存截图
    save_screenshots: Mutex<bool>,
    /// 截图保留天数
    screenshot_retention_days: Mutex<i32>,
    /// 是否允许活跃窗口截图失败后整屏降级
    allow_full_screenshot_fallback: Mutex<bool>,
}

impl CaptureManager {
    /// 创建 CaptureManager 实例
    pub fn new() -> Self {
        Self {
            watcher: Mutex::new(None),
            privacy_guard: PrivacyGuard::new(),
            decision: CaptureDecision::new(),
            state: Mutex::new(CaptureState::default()),
            save_screenshots: Mutex::new(false),
            screenshot_retention_days: Mutex::new(0),
            allow_full_screenshot_fallback: Mutex::new(false),
        }
    }

    // ===================== 捕获控制 =====================

    /// 启动全链路捕获
    pub fn start_capture(&self) -> bool {
        self.privacy_guard.seed_default_rules();
        // 启动 WindowWatcher（如果可用）
        if let Ok(mut guard) = self.watcher.lock() {
            if guard.is_none() {
                let mut watcher = WindowWatcher::new();
                let _ = watcher.start();
                *guard = Some(watcher);
            }
        }
        self.decision.start();
        self.set_base_state(RecordingState::Recording);
        true
    }

    /// 停止捕获
    pub fn stop_capture(&self) -> bool {
        self.decision.stop();
        if let Ok(mut guard) = self.watcher.lock() {
            if let Some(mut watcher) = guard.take() {
                let _ = watcher.stop();
            }
        }
        self.set_base_state(RecordingState::Idle);
        true
    }

    /// 暂停捕获
    pub fn pause_capture(&self) -> bool {
        self.decision.pause();
        self.set_base_state(RecordingState::Paused);
        true
    }

    /// 恢复捕获
    pub fn resume_capture(&self) -> bool {
        self.decision.resume();
        self.set_base_state(RecordingState::Recording);
        true
    }

    /// 获取当前记录状态（含 privacy 覆盖）
    pub fn get_recording_state(&self) -> RecordingState {
        self.state.lock().unwrap().effective_state()
    }

    // ===================== 模块访问 =====================

    /// 获取隐私守卫引用
    pub fn get_privacy_guard(&self) -> &PrivacyGuard {
        &self.privacy_guard
    }

    /// 获取截图决策引用
    pub fn get_capture_decision(&self) -> &CaptureDecision {
        &self.decision
    }

    // ===================== 设置 =====================

    /// 设置是否持久保存截图
    pub fn set_save_screenshots(&self, enabled: bool) {
        *self.save_screenshots.lock().unwrap() = enabled;
    }

    /// 设置截图保留天数
    pub fn set_screenshot_retention_days(&self, days: i32) {
        *self.screenshot_retention_days.lock().unwrap() = days.max(0).min(7);
    }

    /// 设置是否允许活跃窗口截图失败后整屏降级，并立即下发到 CaptureDecision。
    /// 默认 false：窗口截图失败即跳过，绝不自动整屏。
    pub fn set_allow_full_screenshot_fallback(&self, enabled: bool) {
        *self.allow_full_screenshot_fallback.lock().unwrap() = enabled;
        self.decision.set_allow_full_screenshot_fallback(enabled);
    }

    /// 查询当前是否允许整屏降级
    pub fn is_full_screenshot_fallback_allowed(&self) -> bool {
        *self.allow_full_screenshot_fallback.lock().unwrap()
    }

    /// 用户从空闲恢复活动时唤醒决策引擎
    pub fn wake_from_activity(&self, _window_info: &WindowInfo) {
        self.decision.wake_from_activity();
        let mut state = self.state.lock().unwrap();
        if !state.privacy_mode && state.base_state == RecordingState::Idle {
            state.base_state = RecordingState::Recording;
            drop(state);
            self.broadcast_state();
        }
    }

    // ===================== 内部工具 =====================

    /// 设置基础状态并广播
    fn set_base_state(&self, state: RecordingState) {
        {
            let mut s = self.state.lock().unwrap();
            s.base_state = state;
        }
        self.broadcast_state();
    }

    /// 广播状态变化
    fn broadcast_state(&self) {
        let state = self.get_recording_state();
        let state_str = match state {
            RecordingState::Recording => "recording",
            RecordingState::Paused => "paused",
            RecordingState::Idle => "idle",
            RecordingState::Privacy => "privacy",
        };
        EventBus::publish(AppEvent::StateChange {
            state: state_str.to_string(),
        });
    }
}

impl Default for CaptureManager {
    fn default() -> Self {
        Self::new()
    }
}

/// 获取当前时间戳（毫秒）
fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as u64)
        .unwrap_or(0)
}

// ===================== 单例 =====================

/// CaptureManager 单例
static MANAGER_INSTANCE: Lazy<Mutex<Option<CaptureManager>>> = Lazy::new(|| Mutex::new(None));

/// 获取 CaptureManager 单例锁
pub fn get_capture_manager() -> &'static Mutex<Option<CaptureManager>> {
    &MANAGER_INSTANCE
}

/// 初始化单例（app ready 后调用）
pub fn init_capture_manager() {
    let mut guard = MANAGER_INSTANCE.lock().unwrap();
    if guard.is_none() {
        *guard = Some(CaptureManager::new());
    }
}

/// 重置单例（仅供测试）
pub fn reset_capture_manager() {
    let mut guard = MANAGER_INSTANCE.lock().unwrap();
    if let Some(manager) = guard.take() {
        manager.stop_capture();
    }
}

// ===================== 单元测试 =====================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_capture_state_default() {
        let state = CaptureState::default();
        assert_eq!(state.base_state, RecordingState::Idle);
        assert!(!state.privacy_mode);
        assert_eq!(state.effective_state(), RecordingState::Idle);
    }

    #[test]
    fn test_capture_state_privacy_override() {
        let mut state = CaptureState::default();
        state.base_state = RecordingState::Recording;
        state.privacy_mode = true;
        // 隐私模式覆盖基础状态
        assert_eq!(state.effective_state(), RecordingState::Privacy);
    }

    #[test]
    fn test_capture_manager_creation() {
        let manager = CaptureManager::new();
        assert_eq!(manager.get_recording_state(), RecordingState::Idle);
        assert!(!manager.is_full_screenshot_fallback_allowed());
    }

    #[test]
    fn test_set_save_screenshots() {
        let manager = CaptureManager::new();
        manager.set_save_screenshots(true);
        assert!(*manager.save_screenshots.lock().unwrap());
        manager.set_save_screenshots(false);
        assert!(!*manager.save_screenshots.lock().unwrap());
    }

    #[test]
    fn test_set_screenshot_retention_days_clamped() {
        let manager = CaptureManager::new();
        manager.set_screenshot_retention_days(100);
        assert_eq!(*manager.screenshot_retention_days.lock().unwrap(), 7);
        manager.set_screenshot_retention_days(-5);
        assert_eq!(*manager.screenshot_retention_days.lock().unwrap(), 0);
        manager.set_screenshot_retention_days(3);
        assert_eq!(*manager.screenshot_retention_days.lock().unwrap(), 3);
    }

    #[test]
    fn test_set_allow_full_screenshot_fallback() {
        let manager = CaptureManager::new();
        assert!(!manager.is_full_screenshot_fallback_allowed());
        manager.set_allow_full_screenshot_fallback(true);
        assert!(manager.is_full_screenshot_fallback_allowed());
    }

    #[test]
    fn test_pause_and_resume() {
        let manager = CaptureManager::new();
        manager.pause_capture();
        assert_eq!(manager.get_recording_state(), RecordingState::Paused);
        manager.resume_capture();
        assert_eq!(manager.get_recording_state(), RecordingState::Recording);
    }
}
